### ISSUE TITLE
Durable local storage

### ISSUE THREAD

**Issue by pixie79** [GoogleCloudPlatform/kubernetes#598]
Is there a way to pin a pod to a minion?

For example we have some data that is stored on the host disk that is persistent between reboots, as such I need to tell the replication controller that this container should be pinned for example to minion1.

**Comment by erictune** [comment]
Reopening because I closed #1515 

I'n my mind the open issue is for me to more closely document the design decision I made when I closed #1515,  and then produce user document with recommendations on how to handle specific use cases for longer-than-pod-lifetime storage (Brendan's work getting e.g. cassandra example being one such specific case).

**Comment by sylus** [comment]
I would be very keen on user documentation for this. I have been looking at the corresponding issues and still am having a degree of confusion on what the recommended practice is.

**Comment by smarterclayton** [comment]
As a side note, I'm going to attempt to resolve https://github.com/openshift/origin/issues/663 by building on top of Tims volume extension work and deliver a "best effort local volume" that can be reused on a host by knowing a unique secret (and once I get the security context work written up, would be protected by Unix user permissions from other containers).  That would allow folks to have something strictly more useful than a one off volume, but less useful than a networked persistent disk guarantee.   The use case is reusing things like caches, build artifacts, etc.

We can then look in the future to a best effort snapshot mechanism that attempts to fill the volume with the most recent info available (and can ensure backups and snapshots are reasonably up to date).  The guarantees would be much weaker than what this originally proposed, which as Eric noted are difficult to cleanly model.  

The VolumeAttachController mentioned in Marks pull is an example of a mechanism that can do that best effort attempt to "attach" a recent snapshot to a node after a pod is scheduled, but before the volume plugin allows the containers to start (perhaps by indicating the maxWait for a snapshot to be available before starting from scratch).

**Comment by erictune** [comment]
@smarterclayton what is the mechanism that makes a pod that wants the "best effort local volume" likely to be scheduled on the node with the volume it wants?

**Comment by smarterclayton** [comment]
Initially, nothing.  In the future, the volume mount logic on the node attempts to bring the latest snapshot to the current node on start.  In the far far future, a finalizer for best effort local volume attempts to influence the scheduler selection criteria (or doesn't - I don't think best effort that good is worth solving because it's really indistinguishable from network attached storage, which is effectively solved).

----- Original Message -----

**Comment by pwFoo** [comment]
Don't know how kubernetes or docker works internal, but durable local storage should be work with  data only container if native docker is used? 
So maybe data containers and a implementation of volumes-from should work to get local data persistent? Because I'll move to a docker host I need a local persistent storage.

But it seems there is no solution in the near future?

**Comment by thockin** [comment]
I am not against some variant of data containers, but I don't really know
how people are using data containers.

Is the goal just to get a stable/recoverable host dir to write into?  Is
the goal to preload said dir with the contents of a docker container?
Something else?

**Comment by pwFoo** [comment]
Data containers are persistent and reboot save. After read some issues hostdir have some permission problems. Data containers could be a reboot save volume solution and simpler to move to another minion than hostdir data if needed (?)
So data containers could be a better way to handle volumes?

**Comment by markturansky** [comment]
@thockin could you explain for me the difference difference between a data container and a persistent volume?  I'd like to understand this issue better.

**Comment by smarterclayton** [comment]
I discussed this in another issue, but I'd prefer a GC'd directory identified by a unique value that anyone who knows could reuse.  Given proper uid support that content is protected by Unix rules and would satisfy the "I need a dir that is best effort reused per host across multiple pods", as a build cache or scratch dir.  But it would need time based GC after the last reference is allocated.

**Comment by eparis** [comment]
To understand data containers, we need to remember that container != image.

In the docker model, you can make changes to your container's filesystem, stop the container/reboot the host, and then you can restart the container and it will still have those FS changes.

In the kube model, we tend to believe that containers are always launched cleanly from an image and storage/filesystem changes should be done 'outside' of the container.

"Data containers" are much more in the docker thoughts. You can create a container and make some changes to the filesystem in that container.  Docker can then mount the filesystem from one container into another container. And the changes are in the 'data container.'

Its like what we do with volumes, but they do it with containers (and like many things docker it is only really elegant on a single host)

An example of a 'data container' could be for configuration. You could create a container filled with your rsyslog configuration and another container which actually has rsyslog. You launch the rsyslog continer mounting the /etc/ files from the configuration container into the container with the daemon. Now you can update the rsyslog container independently of the config, and the config independently of the binary.

Another example would be a container to save stateful data.  Create a container which just has /var/lib/etcd/.  Now mount that containers /var/lib/etcd/ into your etcd container.  You can update/change the etcd container without worry about the data.  You can also 'save' the data container as an image and docker push/docker pull to get the data onto another host, in case you wanted to migrate the data.

I haven't read this issue, so the following statements are likely worth exactly 0. But in general, I am not a fan of dockers mutability of containers. I like that kubernetes has a clear and rational expectation that running the same command 2 times will give the same results. If we chose to use containers under the covers for some type of data storage, that it fine, but I really hope we don't expose that to the user. We should expose some functionality to the system user, not some underlying detail....

**Comment by pikeas** [comment]
It's been close to a year since this issue was opened, is Kubernetes really no closer to providing durable local storage and nominal/stateful services (#260)?

**Comment by pwFoo** [comment]
Why not create a data volume which creates a data only container in the background?
It should be added to the container / pod by volumes-from docker option and is persistent / durable.

**Comment by thockin** [comment]
Folding this together with #7562.
